llm_studio/src/datasets/text_causal_language_modeling_ds.py

```diff
@@ -304,23 +335,32 @@ def sanity_check(cls, df: pd.DataFrame, cfg: Any, mode: str = "train"):
     def __getitem__(self, idx: int) -> Dict:
         """Reads a single text observation."""
         sample = dict()
-        prompt_encoding, answer_encoding = self._get_prompt_and_answer_encoding(idx)
+        system_encoding, prompt_encoding, answer_encoding = self._get_sample_encoding(
+            idx
+        )
         rlhf_is_in_training_mode = self.cfg.training.use_rlhf and self.mode == "train"
 
         if rlhf_is_in_training_mode:
             # ground truth answer not used in RLHF training
-            encodings = [[prompt_encoding, torch.empty(0)]]
+            encodings = [[system_encoding, prompt_encoding, torch.empty(0)]]
         else:
-            encodings = [[prompt_encoding, answer_encoding]]
+            encodings = [[system_encoding, prompt_encoding, answer_encoding]]
 
         parent_encodings, reward_model_parent_prompt_text = self.get_parent_encodings(
             idx
         )
         encodings = parent_encodings + encodings
 
-        sample["reward_model_prompt_text"] = (
-            reward_model_parent_prompt_text + self.raw_prompts[idx]
-        )
+        # get first system encoding
+        system_encoding = encodings[0][0]
+        # remove system encodings from list of encodings
+        encodings = [encoding[1:] for encoding in encodings]
+        encodings[0][0] = torch.cat([system_encoding, encodings[0][0]])
```

Comment:
I don't fully grasp that part and why we append it already here. Should it be prepended at the end maybe?

Reply:
So currently each sample in `encodings` has the system prompt, but we only need it for the root, which is why we pick it from `encodings[0][0]` and then remove it again, so that we only have prompt/answer as before.
Then we just concat it to the very first prompt and have the typical workflow.

Then we only need to have this extra functionality at the end, to make sure it is prepended if max token length is exceeded. Otherwise we would always need to prepend it manually, which we could also do, but a bit more messy because we would need to do it after the padding, and then we would need to find the index where the padding ends etc.

Bit messy, but "simplest" solution I could think of. Im sure there are better ways.

Reply:
Thank you, now I understand. Makes perfect sense.
Could you maybe add a few short comments in the code for that reasoning. Basically what you said above.